parsec/event_bus.py

```diff
@@ -15,40 +33,54 @@ class MetaEvent(Enum):
     EVENT_DISCONNECTED = "event.disconnected"
 
 
+CustomEvent = TypeVar("CustomEvent", bound=Enum)
+AllEvent = Union[MetaEvent, CustomEvent]
+
+
+class EventCallback(Protocol):
+    def __call__(self, event: Enum, **kwargs: Any) -> None:
+        ...
+
+
+class EventFilterCallback(Protocol):
+    def __call__(self, event: Enum, **kwargs: Any) -> bool:
+        ...
+
+
 class EventWaiter:
-    def __init__(self, filter):
+    def __init__(self, filter: Optional[EventFilterCallback]):
         self._filter = filter
         self._event_occured = trio.Event()
-        self._event_result = None
+        self._event_result: Optional[Tuple[Enum, Dict[str, Any]]] = None
```

Comment:
Maybe this attribute should be splited as:
```python
self._event_result_type: Optional[Enum] = None
self._event_result_values: Dict[str, Any] = {}
```

Also `Enum` is a bit generic.. Is there a base class for event types?

Reply:
> Also Enum is a bit generic.. Is there a base class for event types?

Good catch, it should be an `AllEvent` !

> Maybe this attribute should be splited as:

Not sure: the non-splitted form make it clear the event is present or not, splitting it seems only better because it makes the typing information smaller, which doesn't seems like a very compelling argument to me ;-)

Reply:
> smaller

I'd say "more readable" but alright :)